k-distribution/k-tutorial/1_basic/16_collections/README.md

```diff
@@ -1 +1,344 @@
+`<cell> _:List (Pattern) </cell>`, the meaning of `<cell> Pattern ...</cell>`
+is the same as `<cell> (Pattern) _:List</cell>`. Because list patterns with
+multiple variables of sort `List` are not allowed, it is an error to write
+`<cell>... Pattern ...</cell>`.
+
+Here is an example of a simple definition that implements a very basic
+function-call semantics using a `List` as a function stack (`lesson-16-a.k`):
+
+```k
+module LESSON-16-B-SYNTAX
+  imports INT-SYNTAX
+  import ID-SYNTAX
+
+  syntax Exp ::= Id "(" ")" | Int
+  syntax Stmt ::= "return" Exp ";" [strict]
+  syntax Decl ::= "fun" Id "(" ")" "{" Stmt "}"
+  syntax Pgm ::= List{Decl,""}
+  syntax Id ::= "main" [token]
+endmodule
+
+module LESSON-16-B
+  imports LESSON-16-B-SYNTAX
+  imports BOOL
+
+  configuration <T>
+                  <k> $PGM:Pgm ~> main () </k>
+                  <functions> .Map </functions>
+                  <fstack> .List </fstack>
+                </T>
+
+  // declaration sequence
+  rule <k> D:Decl P:Pgm => D ~> P ...</k>
+  rule <k> .Pgm => . ...</k>
+
+  // function definitions
+  rule <k> fun X:Id () { S } => . ...</k>
+       <functions>... .Map => X |-> S ...</functions>
+
+  // function call
+  syntax KItem ::= stackFrame(K)
+  rule <k> X:Id () ~> K => S </k>
+       <functions>... X |-> S ...</functions>
+       <fstack> .List => ListItem(stackFrame(K)) ...</fstack>
+
+  // return statement
+  rule <k> return I:Int ; ~> _ => I ~> K </k>
+       <fstack> ListItem(stackFrame(K)) => .List ...</fstack>
+
+  syntax Bool ::= isKResult(K) [function, symbol]
+  rule isKResult(_:Int) => true
+  rule isKResult(_) => false [owise]
+endmodule
+```
+
+Notice that we have declared the production `syntax Id ::= "main" [token]`.
+Since we use the `ID-SYNTAX` module, this declaration is necessary in order to
+be able to refer to the `main` identifier directly in the configuration
+declaration. Our `<k>` cell now contains a `K` sequence initially: first we
+process all the declarations in the program, then we call the `main` function.
+
+Consider the following program (`foo.func`):
+
+```
+int foo() { return 5; }
+int main() { return foo(); }
+```
+
+When we `krun` this program, we should get the following output:
+
+```
+<T>
+  <k>
+    5 ~> .
+  </k>
+  <functions>
+    foo |-> return 5 ;
+    main |-> return foo ( ) ;
+  </functions>
+  <fstack>
+    .List
+  </fstack>
+</T> 
+```
+
+Note that we have successfully put on the `<k>` cell the value returned by the
+`main` function.
+
+### Exercise
+
+Add a term of sort `Id` to the `stackFrame` operator to keep track of the
+name of the function in that stack frame. Then write a function
+`syntax String ::= printStackTrace(List)` that takes the contents of the
+`<fstack>` cell and pretty prints the current stack trace. Test this function
+by creating a new expression that returns the current stack trace as a string.
+Make sure to update `isKResult` and the `Exp` sort as appropriate to allow
+strings as values.
```

Comment:
Tell them teh module name in `domains.md` to find the string operators (eg. `_+String_`) and to find the hook they need (I think `Id2String`).